src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -113,41 +120,34 @@ private void SetEntry(CacheEntry entry)
             // Initialize the last access timestamp at the time the entry is added
             entry.LastAccessed = utcNow;
 
-            var added = false;
             CacheEntry priorEntry;
-
-            _entryLock.EnterWriteLock();
-            try
+            if (_entries.TryRemove(entry.Key, out priorEntry))
```

Comment:
I depends on what we think is correct. Before MemoryCache was blocking if there was write operation on the way so you were guaranteed to get latest value. In current state of this PR you can get `null` in some cases, in change proposed by @JunTaoLuo you would get old value instead null. I don't think we can achieve exactly same behavior that `ReaderWriterLock` had so we need to decide how much is enough.